solutions/CSES Fixed-Length Paths I.mdx

```diff
@@ -0,0 +1,385 @@
+in the centroid tree takes amortized $\mathcal O(N)$ time to process, the total complexity is $\mathcal O(N log N)$.
+
+## Code
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+typedef long long ll;
+using namespace std;
+
+int n, k;
+vector<int> graph[200001];
+int subtree[200001];
+
+ll ans = 0;
+int cnt[200001]{1}, mx_depth;
+bool processed[200001];
+
+int get_subtree_sizes(int node, int parent = 0) {
+    subtree[node] = 1;
+    for (int i : graph[node]) if (!processed[i] && i != parent)
+        subtree[node] += get_subtree_sizes(i, node);
+    return subtree[node];
+}
+
+int get_centroid(int desired, int node, int parent = 0) {
+    for (int i : graph[node])
+        if (!processed[i] && i != parent && subtree[i] >= desired)
+            return get_centroid(desired, i, node);
+    return node;
+}
+
+void get_cnt(int node, int parent, bool filling, int depth = 1) {
+    if (depth > k) return;
+    mx_depth = max(mx_depth, depth);
+    if (filling) cnt[depth]++;
+    else ans += cnt[k - depth];
+    for (int i : graph[node]) if (!processed[i] && i != parent)
+        get_cnt(i, node, filling, depth + 1);
+}
+
+void centroid_decomp(int node = 1) {
+    int centroid = get_centroid(get_subtree_sizes(node) >> 1, node);
+    processed[centroid] = true;
+    mx_depth = 0;
+    for (int i : graph[centroid]) if (!processed[i]) {
+        get_cnt(i, centroid, false);
+        get_cnt(i, centroid, true);
+    }
+    fill(cnt + 1, cnt + mx_depth + 1, 0);
+    for (int i : graph[centroid]) if (!processed[i]) centroid_decomp(i);
+}
+
+int main() {
+    cin.tie(0)->sync_with_stdio(0);
+    cin >> n >> k;
+    for (int i = 1; i < n; i++) {
+        int u, v;
+        cin >> u >> v;
+        graph[u].push_back(v);
+        graph[v].push_back(u);
+    }
+    centroid_decomp();
+    cout << ans;
+    return 0;
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+<Info title="Bonus Solution">
+
+It's also possible to solve this problem in linear time!
+
+We use the same idea as the above solution but with prefix sums instead of a BIT and with [small-to-large merging](/plat/merging) in a single DFS instead of centroid decomposition. This is because we can merge two prefix sum arrays $A$ and $B$ in $\mathcal O(\min(|A|, |B|))$ time.
+
+If we let $d_i$ denote the maximum depth in node $i$'s subtree, then the resulting time complexity is
+$$
+\mathcal O\left(\sum_{i = 1}^N\left(d_i - \max_{c \in i \text{'s children}}(d_c)\right)\right).
+$$
+Since $d_i = \max_{c \in i \text{'s children}}(d_c) + 1$, the above expression simplifies to $\mathcal O(N)$.
+
+Below is Ben's code implementing this idea:
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+using ll = long long;
+using db = long double; // or double, if TL is tight
+using str = string; // yay python!
+ 
+using pi = pair<int,int>;
```

Comment:
un-macro/un-templatify this code?

Reply:
Idk I personally think this is fine since it's not excessive, but I'm biased towards templates :P